### ISSUE TITLE
valgrind errors in t/op/vec.t

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13638]
Migrated from [rt.perl.org#121366](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121366) (status was 'resolved')

Searchable as RT121366$

**Comment by p5pRT** [comment]
### From @khwilliamson

A bisect yields this&#8203;:

032061d233a4bb16c1677ef64615bdb15de5b8a1 is the first bad commit
 commit 032061d233a4bb16c1677ef64615bdb15de5b8a1
 Author&#8203;: Brian Fraser \<fraserbn@&#8203;gmail\.com>
 Date&#8203;:   Sat May 26 17&#8203;:35&#8203;:26 2012 \-0300
 
&nbsp;     Fix for \[perl \#9423\] vec assignments generate 2 warnings
&nbsp;     
&nbsp;     Before this commit\, this&#8203;:
&nbsp;     
&nbsp;         vec\(my $v\,0\,1\) = 1;
&nbsp;     
&nbsp;     would've produced four warnings about uninitialized values;
&nbsp;     however\, the ticket argued that these were spurious\.
&nbsp;     
&nbsp;     This commit removes the warning in the case of lvalue vec\, since it is
&nbsp;     similar to |=\, but leaves it in place for rvalue vec\.
 
 :100644 100644 1bd16b5e668fdbd0d0be725e26625fd93cbf4051 b2b6546d8cd32bfffa1c65112697a04de9560090 M      doop\.c
 :040000 040000 febb4034f8c7bc380a93193e5df0039f1e16c4d2 88120f8bce08edef4a05927b045a3a465087a857 M      t
 bisect run success
 That took 793 seconds
\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Mar 03 11&#8203;:28&#8203;:21 2014\, khw wrote&#8203;:
> A bisect yields this&#8203;:
> 
> 032061d233a4bb16c1677ef64615bdb15de5b8a1 is the first bad commit
>  commit 032061d233a4bb16c1677ef64615bdb15de5b8a1
>  Author&#8203;: Brian Fraser \<fraserbn@&#8203;gmail\.com>
>  Date&#8203;:   Sat May 26 17&#8203;:35&#8203;:26 2012 \-0300

This is happening when using vec\(\) to modify an undef sv\, eg&#8203;:

$ valgrind \-q \./perl \-e 'vec\($Foo\, 0\, 1\) = 1'
==814== Conditional jump or move depends on uninitialised value\(s\)
==814==    at 0x5190C7&#8203;: Perl\_sv\_pvn\_force\_flags \(sv\.c&#8203;:9505\)
==814==    by 0x580C9B&#8203;: Perl\_do\_vecset \(doop\.c&#8203;:927\)
==814==    by 0x4D30AD&#8203;: Perl\_magic\_setvec \(mg\.c&#8203;:2305\)
==814==    by 0x4CE522&#8203;: Perl\_mg\_set \(mg\.c&#8203;:279\)
==814==    by 0x4EB417&#8203;: Perl\_pp\_sassign \(pp\_hot\.c&#8203;:223\)
==814==    by 0x4C42C9&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2425\)
==814==    by 0x44BEC4&#8203;: perl\_run \(perl\.c&#8203;:2449\)
==814==    by 0x422934&#8203;: main \(perlmain\.c&#8203;:112\)
==814== 

The error is on this line&#8203;:

>>>    if \(s \!= SvPVX\_const\(sv\)\) \{    /\* Almost\, but not quite\, sv\_setpvn\(\) \*/
&nbsp;        if \(SvROK\(sv\)\)
&nbsp;        sv\_unref\(sv\);

where we examine sv\->sv\_u\.svu\_pv when it hasn't been initialized yet\.

The following change prevents the error and passes all tests \(DEBUGGING\)&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/sv.c b/sv.c
index dcb1d5e..4742c7d 100644
--- a/sv.c
+++ b/sv.c
@@ -9502,7 +9502,8 @@ Perl_sv_pvn_force_flags(pTHX_ SV *const sv, STRLEN *const 
        if (lp)
            *lp = len;
 
-       if (s != SvPVX_const(sv)) {     /* Almost, but not quite, sv_setpvn() */
+       if (SvTYPE(sv) < SVt_PV ||
+            s != SvPVX_const(sv)) {    /* Almost, but not quite, sv_setpvn() */
            if (SvROK(sv))
                sv_unref(sv);
            SvUPGRADE(sv, SVt_PV);              /* Never FALSE */

Tony
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @shlomif

On Sun Mar 23 21&#8203;:15&#8203;:44 2014\, tonyc wrote&#8203;:
> On Mon Mar 03 11&#8203;:28&#8203;:21 2014\, khw wrote&#8203;:
> > A bisect yields this&#8203;:
> >
> > 032061d233a4bb16c1677ef64615bdb15de5b8a1 is the first bad commit
> >  commit 032061d233a4bb16c1677ef64615bdb15de5b8a1
> >  Author&#8203;: Brian Fraser \<fraserbn@&#8203;gmail\.com>
> >  Date&#8203;:   Sat May 26 17&#8203;:35&#8203;:26 2012 \-0300
> 
> This is happening when using vec\(\) to modify an undef sv\, eg&#8203;:
> 
> $ valgrind \-q \./perl \-e 'vec\($Foo\, 0\, 1\) = 1'
> ==814== Conditional jump or move depends on uninitialised value\(s\)
> ==814==    at 0x5190C7&#8203;: Perl\_sv\_pvn\_force\_flags \(sv\.c&#8203;:9505\)
> ==814==    by 0x580C9B&#8203;: Perl\_do\_vecset \(doop\.c&#8203;:927\)
> ==814==    by 0x4D30AD&#8203;: Perl\_magic\_setvec \(mg\.c&#8203;:2305\)
> ==814==    by 0x4CE522&#8203;: Perl\_mg\_set \(mg\.c&#8203;:279\)
> ==814==    by 0x4EB417&#8203;: Perl\_pp\_sassign \(pp\_hot\.c&#8203;:223\)
> ==814==    by 0x4C42C9&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2425\)
> ==814==    by 0x44BEC4&#8203;: perl\_run \(perl\.c&#8203;:2449\)
> ==814==    by 0x422934&#8203;: main \(perlmain\.c&#8203;:112\)
>  ==814==
> 
> The error is on this line&#8203;:
> 
> >>> if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\,
> >>> sv\_setpvn\(\) \*/
> if \(SvROK\(sv\)\)
>     sv\_unref\(sv\);
> 
> where we examine sv\->sv\_u\.svu\_pv when it hasn't been initialized yet\.
> 
> The following change prevents the error and passes all tests
> \(DEBUGGING\)&#8203;:
> 
> diff \-\-git a/sv\.c b/sv\.c
> index dcb1d5e\.\.4742c7d 100644
> \-\-\- a/sv\.c
> \+\+\+ b/sv\.c
>  @&#8203;@&#8203; \-9502\,7 \+9502\,8 @&#8203;@&#8203; Perl\_sv\_pvn\_force\_flags\(pTHX\_ SV \*const sv\,
> STRLEN \*const
>         if \(lp\)
>             \*lp = len;
> 
> \-       if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\,
> sv\_setpvn\(\) \*/
> \+       if \(SvTYPE\(sv\) \< SVt\_PV ||
> \+            s \!= SvPVX\_const\(sv\)\) \{    /\* Almost\, but not quite\,
> sv\_setpvn\(\) \*/
>             if \(SvROK\(sv\)\)
>                 sv\_unref\(sv\);
>             SvUPGRADE\(sv\, SVt\_PV\);              /\* Never FALSE \*/
> 

I wanted to test your patch but could not because it didn't apply cleanly using "patch \-p1 \< …"\. The problem seems to be that it is all whitespace\, while the sources use a mixture of tabs and spaces\. I can apply it manually and provide a new patch\, but shouldn't the patch be attached?

Regards\,

\-\- Shlomi Fish

**Comment by p5pRT** [comment]
### From @shlomif

On Tue Mar 25 01&#8203;:24&#8203;:32 2014\, shlomif wrote&#8203;:
> On Sun Mar 23 21&#8203;:15&#8203;:44 2014\, tonyc wrote&#8203;:
> > On Mon Mar 03 11&#8203;:28&#8203;:21 2014\, khw wrote&#8203;:
> > > A bisect yields this&#8203;:
> > >
> > > 032061d233a4bb16c1677ef64615bdb15de5b8a1 is the first bad commit
> > >  commit 032061d233a4bb16c1677ef64615bdb15de5b8a1
> > >  Author&#8203;: Brian Fraser \<fraserbn@&#8203;gmail\.com>
> > >  Date&#8203;:   Sat May 26 17&#8203;:35&#8203;:26 2012 \-0300
> >
> > This is happening when using vec\(\) to modify an undef sv\, eg&#8203;:
> >
> > $ valgrind \-q \./perl \-e 'vec\($Foo\, 0\, 1\) = 1'
> > ==814== Conditional jump or move depends on uninitialised value\(s\)
> > ==814==    at 0x5190C7&#8203;: Perl\_sv\_pvn\_force\_flags \(sv\.c&#8203;:9505\)
> > ==814==    by 0x580C9B&#8203;: Perl\_do\_vecset \(doop\.c&#8203;:927\)
> > ==814==    by 0x4D30AD&#8203;: Perl\_magic\_setvec \(mg\.c&#8203;:2305\)
> > ==814==    by 0x4CE522&#8203;: Perl\_mg\_set \(mg\.c&#8203;:279\)
> > ==814==    by 0x4EB417&#8203;: Perl\_pp\_sassign \(pp\_hot\.c&#8203;:223\)
> > ==814==    by 0x4C42C9&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2425\)
> > ==814==    by 0x44BEC4&#8203;: perl\_run \(perl\.c&#8203;:2449\)
> > ==814==    by 0x422934&#8203;: main \(perlmain\.c&#8203;:112\)
> >  ==814==
> >
> > The error is on this line&#8203;:
> >
> > >>> if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\,
> > >>> sv\_setpvn\(\) \*/
> > if \(SvROK\(sv\)\)
> >     sv\_unref\(sv\);
> >
> > where we examine sv\->sv\_u\.svu\_pv when it hasn't been initialized yet\.
> >
> > The following change prevents the error and passes all tests
> > \(DEBUGGING\)&#8203;:
> >
> > diff \-\-git a/sv\.c b/sv\.c
> > index dcb1d5e\.\.4742c7d 100644
> > \-\-\- a/sv\.c
> > \+\+\+ b/sv\.c
> >  @&#8203;@&#8203; \-9502\,7 \+9502\,8 @&#8203;@&#8203; Perl\_sv\_pvn\_force\_flags\(pTHX\_ SV \*const sv\,
> > STRLEN \*const
> >         if \(lp\)
> >             \*lp = len;
> >
> > \-       if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\,
> > sv\_setpvn\(\) \*/
> > \+       if \(SvTYPE\(sv\) \< SVt\_PV ||
> > \+            s \!= SvPVX\_const\(sv\)\) \{    /\* Almost\, but not quite\,
> > sv\_setpvn\(\) \*/
> >             if \(SvROK\(sv\)\)
> >                 sv\_unref\(sv\);
> >             SvUPGRADE\(sv\, SVt\_PV\);              /\* Never FALSE \*/
> >
> 
> I wanted to test your patch but could not because it didn't apply
> cleanly using "patch \-p1 \< …"\. The problem seems to be that it is all
> whitespace\, while the sources use a mixture of tabs and spaces\. I can
> apply it manually and provide a new patch\, but shouldn't the patch be
> attached?

The reworked patch is attached to this comment\. All tests are successful here \(Mageia Linux x86\-64 5/Cauldron\)\.

Regards\,

\-\- Shlomi Fish

**Comment by p5pRT** [comment]
### From @shlomif

<details><summary>rt121366.diff</summary>

```diff
diff --git a/sv.c b/sv.c
index dcb1d5e..9afe983 100644
--- a/sv.c
+++ b/sv.c
@@ -9502,7 +9502,8 @@ Perl_sv_pvn_force_flags(pTHX_ SV *const sv, STRLEN *const lp, const I32 flags)
 	if (lp)
 	    *lp = len;
 
-	if (s != SvPVX_const(sv)) {	/* Almost, but not quite, sv_setpvn() */
+	if (SvTYPE(sv) < SVt_PV ||
+	    s != SvPVX_const(sv)) {	/* Almost, but not quite, sv_setpvn() */
 	    if (SvROK(sv))
 		sv_unref(sv);
 	    SvUPGRADE(sv, SVt_PV);		/* Never FALSE */

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

On Mon\, Mar 24\, 2014 at 5&#8203;:15 AM\, Tony Cook via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Mon Mar 03 11&#8203;:28&#8203;:21 2014\, khw wrote&#8203;:
>> A bisect yields this&#8203;:
>>
>> 032061d233a4bb16c1677ef64615bdb15de5b8a1 is the first bad commit
>>  commit 032061d233a4bb16c1677ef64615bdb15de5b8a1
>>  Author&#8203;: Brian Fraser \<fraserbn@&#8203;gmail\.com>
>>  Date&#8203;:   Sat May 26 17&#8203;:35&#8203;:26 2012 \-0300
>
> This is happening when using vec\(\) to modify an undef sv\, eg&#8203;:
>
> $ valgrind \-q \./perl \-e 'vec\($Foo\, 0\, 1\) = 1'
> ==814== Conditional jump or move depends on uninitialised value\(s\)
> ==814==    at 0x5190C7&#8203;: Perl\_sv\_pvn\_force\_flags \(sv\.c&#8203;:9505\)
> ==814==    by 0x580C9B&#8203;: Perl\_do\_vecset \(doop\.c&#8203;:927\)
> ==814==    by 0x4D30AD&#8203;: Perl\_magic\_setvec \(mg\.c&#8203;:2305\)
> ==814==    by 0x4CE522&#8203;: Perl\_mg\_set \(mg\.c&#8203;:279\)
> ==814==    by 0x4EB417&#8203;: Perl\_pp\_sassign \(pp\_hot\.c&#8203;:223\)
> ==814==    by 0x4C42C9&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2425\)
> ==814==    by 0x44BEC4&#8203;: perl\_run \(perl\.c&#8203;:2449\)
> ==814==    by 0x422934&#8203;: main \(perlmain\.c&#8203;:112\)
> ==814==
>
> The error is on this line&#8203;:
>
>>>>     if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\, sv\_setpvn\(\) \*/
>             if \(SvROK\(sv\)\)
>                 sv\_unref\(sv\);
>
> where we examine sv\->sv\_u\.svu\_pv when it hasn't been initialized yet\.
>
> The following change prevents the error and passes all tests \(DEBUGGING\)&#8203;:
>
> diff \-\-git a/sv\.c b/sv\.c
> index dcb1d5e\.\.4742c7d 100644
> \-\-\- a/sv\.c
> \+\+\+ b/sv\.c
> @&#8203;@&#8203; \-9502\,7 \+9502\,8 @&#8203;@&#8203; Perl\_sv\_pvn\_force\_flags\(pTHX\_ SV \*const sv\, STRLEN \*const
>         if \(lp\)
>             \*lp = len;
>
> \-       if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\, sv\_setpvn\(\) \*/
> \+       if \(SvTYPE\(sv\) \< SVt\_PV ||
> \+            s \!= SvPVX\_const\(sv\)\) \{    /\* Almost\, but not quite\, sv\_setpvn\(\) \*/
>             if \(SvROK\(sv\)\)
>                 sv\_unref\(sv\);
>             SvUPGRADE\(sv\, SVt\_PV\);              /\* Never FALSE \*/
>
> Tony
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: new
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121366

Thanks for picking up this\, Tony\! I haven't been able to get valgrind
or gdb to work on my Mac\, so I've been stuck being useless for a
while\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Mar 23 21&#8203;:15&#8203;:44 2014\, tonyc wrote&#8203;:
> This is happening when using vec\(\) to modify an undef sv\, eg&#8203;:
> 
> $ valgrind \-q \./perl \-e 'vec\($Foo\, 0\, 1\) = 1'
> ==814== Conditional jump or move depends on uninitialised value\(s\)
> ==814==    at 0x5190C7&#8203;: Perl\_sv\_pvn\_force\_flags \(sv\.c&#8203;:9505\)
> ==814==    by 0x580C9B&#8203;: Perl\_do\_vecset \(doop\.c&#8203;:927\)
> ==814==    by 0x4D30AD&#8203;: Perl\_magic\_setvec \(mg\.c&#8203;:2305\)
> ==814==    by 0x4CE522&#8203;: Perl\_mg\_set \(mg\.c&#8203;:279\)
> ==814==    by 0x4EB417&#8203;: Perl\_pp\_sassign \(pp\_hot\.c&#8203;:223\)
> ==814==    by 0x4C42C9&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:2425\)
> ==814==    by 0x44BEC4&#8203;: perl\_run \(perl\.c&#8203;:2449\)
> ==814==    by 0x422934&#8203;: main \(perlmain\.c&#8203;:112\)
>  ==814==
> 
> The error is on this line&#8203;:
> 
> >>> if \(s \!= SvPVX\_const\(sv\)\) \{     /\* Almost\, but not quite\,
> >>> sv\_setpvn\(\) \*/
> if \(SvROK\(sv\)\)
>     sv\_unref\(sv\);
> 
> where we examine sv\->sv\_u\.svu\_pv when it hasn't been initialized yet\.

Pushed the fix as e14119056960dbe28537c9870667b5b920d9d731\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
